indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java

```diff
@@ -452,6 +454,66 @@ public boolean isPerfectRollup()
     return tuningConfig != null && tuningConfig.isForceGuaranteedRollup();
   }
 
+  /**
+   * Returns true even when granularitySpec is not known since rollup is then decided based on segment analysis.
+   */
+  static boolean isPossiblyRollup(@Nullable ClientCompactionTaskGranularitySpec granularitySpec)
+  {
+    return !(granularitySpec != null && Boolean.FALSE.equals(granularitySpec.isRollup()));
+  }
+
+  static boolean isRangePartitioned(@Nullable TuningConfig tuningConfig)
+  {
+    return tuningConfig != null && tuningConfig.getPartitionsSpec() instanceof DimensionRangePartitionsSpec;
+  }
+
+  static boolean isRollupOnStringDimension(
+      DimensionsSpec dimensionsSpec,
+      @Nullable ClientCompactionTaskGranularitySpec granularitySpec
+  )
+  {
+    if (!isPossiblyRollup(granularitySpec)) {
+      return false;
+    }
+    return dimensionsSpec.getDimensions()
+                         .stream()
+                         .anyMatch(dim -> ColumnType.STRING.equals(dim.getColumnType()));
+  }
+
+  static boolean isPartitionedOnStringDimension(DimensionsSpec dimensionsSpec, @Nullable TuningConfig tuningConfig)
+  {
+    if (!isRangePartitioned(tuningConfig)) {
+      return false;
+    }
+    DimensionRangePartitionsSpec partitionsSpec = (DimensionRangePartitionsSpec) tuningConfig.getPartitionsSpec();
+    return dimensionsSpec.getDimensions()
+                         .stream()
+                         .anyMatch(
+                             dim -> ColumnType.STRING.equals(dim.getColumnType())
+                                    && partitionsSpec.getPartitionDimensions().contains(dim.getName())
+                         );
+  }
+
+  /**
+   * In MSQ, MVDs need conversion to array for rollup, and they aren't supported as partition columns. This check
+   * therefore is particularly for MSQCompactionRunner; it returns true when dimension types aren't known
+   * from CompactionTask spec, or if either rollup or partition dimensions contain any string-type column.
+   */
+  public static boolean needMultiValuedDimensions(CompactionTask compactionTask)
```

Comment:
No point making this method static if it requires passing the `CompactionTask` into it.
Also, instead of making this method public or package private, `CompactionTask` should instead expose a method `needsToDownloadSegments()` or something. That can internally call this one.

Reply:
> No point making this method static if it requires passing the CompactionTask into it.

Agree. One possibility is that its value can be assigned to a final field and the getter is exposed

> Also, instead of making this method public or package private, CompactionTask should instead expose a method needsToDownloadSegments() or something. That can internally call this one.

This would need some refactoring since the effective granularitySpec on which download depends is computed inside `createDataSchemaForIntervals`, just before passing this computed value to `ExistingSegmentAnalyzer` where the final need to download is evaluated. Also, each `need*` flag is separately communicated to the Analyzer class to do selective processing. 

Does the proposed approach of assigning `needMultiValuedDimensions` to a private field and creating tests on it work? In fact, all such fields apart from `needGranularitySpec` can be populated in CompactionTask constructor and passed to the Analyzer

Reply:
Just relooked at code. Only segmentGranularity is computed later, so all need* flags and an additional new needsDownload flag can be computed in the CompactionTask constructor itself. The needsDownload can have a public getter for tests.

Reply:
If only needed in test, keep the method as package private rather than public.

Reply:
public is needed since the tests reside in the MSQ extension